src/twisted/web/http.py

```
@@ -1825,75 +1823,97 @@ class _ChunkedTransferDecoder:
-            if self.length < 0:
-                raise _MalformedChunkedDataError("Chunk-size must not be negative.")
-            elif self.length == 0:
-                self.state = "TRAILER"
-            else:
-                self.state = "BODY"
-            return rest
+    def _dataReceived_CHUNK_LENGTH(self) -> bool:
+        """
+        Read the chunk size line, ignoring any extensions.
+
+        @returns: C{True} once the line has been read and removed from
+            C{self._buffer}.  C{False} when more data is required.
+
+        @raises _MalformedChunkedDataError: when the chunk size cannot be
+            decoded.
+        """
+        eolIndex = self._buffer.find(b"\r\n", self._start)
+        if eolIndex == -1:
+            self._start = len(self._buffer) - 1
+            return False
+
+        endOfLengthIndex = self._buffer.find(b";", 0, eolIndex)
+        if endOfLengthIndex == -1:
+            endOfLengthIndex = eolIndex
+        try:
+            length = int(self._buffer[0:endOfLengthIndex], 16)
+        except ValueError:
+            raise _MalformedChunkedDataError("Chunk-size must be an integer.")
+
+        if length < 0:
+            raise _MalformedChunkedDataError("Chunk-size must not be negative.")
+        elif length == 0:
+            self.state = "TRAILER"
         else:
-            self._buffer = data
-            return b""
+            self.state = "BODY"
+
+        self.length = length
+        del self._buffer[0 : eolIndex + 2]
+        self._start = 0
+        return True
 
-    def _dataReceived_CRLF(self, data):
-        if data.startswith(b"\r\n"):
+    def _dataReceived_CRLF(self) -> bool:
+        if self._buffer.startswith(b"\r\n"):
             self.state = "CHUNK_LENGTH"
-            return data[2:]
+            del self._buffer[0:2]
+            return True
         else:
-            self._buffer = data
-            return b""
+            return False
```

Maybe remove else and reduce indentation.

A similar comment to `TRAILER` if self._buffer has size 1 or 2 and don't start with `\r` or \r\n` maybe fail fast and receiving more data make no sense.

I see that this is the same behaviour as trunk... so it's ok to have it fixed in a separate ticket.

---

Thanks for pointing this out. I filed https://twistedmatrix.com/trac/ticket/10137 and added some comments referencing it.

There is a potential similar issue in the chunk length line. As I was tinkering with it I wondered if there should be a length limit.